Use Java 8 Time API for sql datetime types

geraint0923 opened the pull request:
Use Java 8 Time APT for SqlDate, SqlTime, SqlTimeWithTimeZone,
SqlTimestamp, and SqlTimestampWithTimeZone.

This will fix #5057

electrum commented:
This needs tests which cover the changed behavior for old dates.

geraint0923 commented:
@electrum I will add the tests soon.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimestamp.java:
```diff
@@ -66,8 +66,7 @@ public boolean equals(Object obj)
-        format.setTimeZone(getTimeZoneForKey(sessionTimeZoneKey));
-        return format.format(new Date(millisUtc));
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS");
+        return ZonedDateTime.ofInstant(Instant.ofEpochMilli(millisUtc), ZoneId.of(sessionTimeZoneKey.getId())).format(formatter);
```
`Instant.ofEpochMilli(...).atZone(...)` instead of `ZonedDateTime.ofInstant(Instant.ofEpochMilli(...), ...)`

haozhun replied:
Initialize the `formatter` statically.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlTimestampWithTimeZone.java:
```diff
@@ -81,8 +82,7 @@ public boolean equals(Object obj)
-        format.setTimeZone(getTimeZoneForKey(timeZoneKey));
-        return format.format(new Date(millisUtc)) + " " + timeZoneKey.getId();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS VV");
+        return ZonedDateTime.ofInstant(Instant.ofEpochMilli(millisUtc), ZoneId.of(timeZoneKey.getId())).format(formatter);
```
`Instant.ofEpochMilli(...).atZone(...)` instead of `ZonedDateTime.ofInstant(Instant.ofEpochMilli(...), ...)`

haozhun replied:
Initialize the `formatter` statically.

haozhun commented on presto-spi/src/main/java/com/facebook/presto/spi/type/SqlDate.java:
```diff
@@ -60,8 +57,6 @@ public boolean equals(Object obj)
-        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
-        format.setTimeZone(getTimeZoneForKey(UTC_KEY));
-        return format.format(new Date(TimeUnit.DAYS.toMillis(days)));
+        return LocalDate.ofEpochDay(days).format(DateTimeFormatter.ofPattern("YYYY-MM-dd").withZone(ZoneId.of("UTC")));
```
Calling `withZone` when formatting a temporal object that does not contain an instant is no-op unless you are printing out the offset. Quote from java doc:

> If the temporal object does not contain an instant, but does contain an offset then an additional check is made. If the normalized override zone is an offset that differs from the offset of the temporal, then a DateTimeException is thrown. In all other cases, the override zone is added to the temporal, replacing any previous zone, but without changing the date/time.

haozhun replied:
`toString` suffices in this case. You don't need a formatter.

geraint0923 commented:
@haozhun Addressed your comments and add tests.

Since the existing `assert` methods will convert the results to  Java Sql `Date`, `DateTime` when doing comparison or directly compare the internal millisecond or days which will not invoke the modified `toString` function, I add one method to force to invoke `toString` when comparing the results.

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -820,11 +820,51 @@ public void testLocale()
+    public void testDateTimeOutputString()
+    {
+        // SqlDate
+        assertFunctionString("date '2012-12-31'", DateType.DATE, "2012-12-31");
```
static import

geraint0923 replied:
There have been some `static final` variables having the same names as `TIME`, `TIMESTAMP`, and `DATE`.

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -820,11 +820,51 @@ public void testLocale()
+        assertFunctionString("date '1560-04-29'", DateType.DATE, "1560-04-29");
+
+        // SqlTime
+        assertFunctionString("time '00:00:00'", TimeType.TIME, "00:00:00.000");
```
static import

haozhun commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestDateTimeFunctions.java:
```diff
@@ -820,11 +820,51 @@ public void testLocale()
+        assertFunctionString(WEIRD_TIME_LITERAL, TIME_WITH_TIME_ZONE, "03:04:05.321 +07:09");
+
+        // SqlTimestamp
+        assertFunctionString("timestamp '0000-01-02 01:02:03'", TimestampType.TIMESTAMP, "0000-01-02 01:02:03.000");
```
static import

haozhun commented:
Commit message should say something like "Fix toString for sql datetime types" instead.

Make sure your tests pass when they are run in a negative time zone as well.

Otherwise looks good.
